staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go

```diff
@@ -92,6 +92,29 @@ type Extensions struct {
 	//      - type: string
 	//    - ... zero or more
 	XIntOrString bool
+
+	// x-kubernetes-list-map-keys annotates map lists by specifying the keys used
+	// as the index of the map.
+	//
+	// This tag MUST only be used on lists that have the "x-kubernetes-list-type"
+	// extension set to "map". Also, the values specified for this attribute must
+	// be a scalar typed field of the child structure (no nesting is supported).
+	XListMapKeys []string
+
+	// x-kubernetes-list-type annotates a list to further describe its topology.
+	// This extension must only be used on lists and may have 3 possible values:
+	//
+	// 1) `atomic`: the list is treated as a single entity, like a scalar.
+	//      Atomic lists will be entirely replaced when updated. This extension
+	//      may be used on any type of list (struct, scalar, ...).
+	// 2) `set`:
+	//      Sets are lists that must not have multiple times the same value. Each
+	//      value must be a scalar (or another atomic type).
+	// 3) `map`:
+	//      These lists are like maps in that their elements have a non-index key
+	//      used to identify them. Order is preserved upon merge. The map tag
+	//      must only be used on a list with struct elements.
+	XListType *string
```

Comment:
missing validation of these 3 strings.

Reply:
Yes, please. This must be explicit in pkg/apis/apiexensions/validation.go and its tests. It is the place where we reason about our API surface. We cannot leak anything from deep inside a library.